File: sol/solution/0001-0100/0033. Search in Rotated Sorted Array/Solution.scala
object Solution {
    def search(nums: Array[Int], target: Int): Int = {
        var l = 0
        var r = nums.length - 1
        while (l <= r) {
            val mid = (l + r) / 2
            if (nums(mid) == target) return mid
            if (nums(l) <= nums(mid)) {
                if (nums(l) <= target && target < nums(mid)) r = mid - 1
                else l = mid + 1
            } else {
                if (nums(mid) < target && target <= nums(r)) l = mid + 1
                else r = mid - 1
            }
        }
        -1
    }
}

File: sol/solution/0001-0100/0033. Search in Rotated Sorted Array/Solution.rs
impl Solution {
    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        let (mut l, mut r) = (0, nums.len() as i32 - 1);
        while l <= r {
            let mid = l + (r - l) / 2;
            if nums[mid as usize] == target { return mid; }
            if nums[l as usize] <= nums[mid as usize] {
                if nums[l as usize] <= target && target < nums[mid as usize] {
                    r = mid - 1;
                } else {
                    l = mid + 1;
                }
            } else {
                if nums[mid as usize] < target && target <= nums[r as usize] {
                    l = mid + 1;
                } else {
                    r = mid - 1;
                }
            }
        }
        -1
    }
}